JetStreamCompose/jetstream/src/main/java/com/google/jetstream/data/repositories/CachedDataReader.kt

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.jetstream.data.repositories
+
+import com.google.jetstream.data.entities.Movie
+import com.google.jetstream.data.models.MovieCastResponseItem
+import com.google.jetstream.data.models.MovieCategoriesResponseItem
+import com.google.jetstream.data.models.MoviesResponseItem
+import com.google.jetstream.data.util.AssetsReader
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+import kotlinx.serialization.json.Json
+
+internal class CachedDataReader<T>(private val reader: suspend () -> List<T>) {
+    private val mutex = Mutex()
+    private lateinit var cache: List<T>
+
+    suspend fun read(): List<T> {
+        mutex.withLock {
```

Comment:
Interesting. Could you please explain me the purpose of this cache and mutex? In Kotlin, the regular, immutable List is inherently thread-safe for read-only operations. Once it's created, its state cannot be changed, so multiple threads can read from it safely without any additional synchronization.

Reply:
I would like avoid the two+ threads try to initialize the cache of the same reader, e.g. two threads simultaneously access the reader before initializing cache.